exa/core/dao.py

```diff
@@ -0,0 +1,287 @@
+from traitlets import default, validate, TraitError
+import numpy as np
+import pandas as pd
+
+from exa.core.data import Data
+
+
+_CHUNKSIZE = 50000
+
+_op_map = {
+    k.rstrip('_'): getattr(operator, k) for k in [
+        'lt', 'le', 'eq',
+        'ne', 'ge', 'gt',
+        'not_', 'and_', 'or_'
+    ]
+}
+
+_raw_op_map = {
+    'lt': '<', 'le': '<=', 'eq': '=',
+    'ne': '!=', 'ge': '>=', 'gt': '>',
+    'not': 'not', 'and': 'and', 'or': 'or'
+}
+
+class SQLInjectionError(Exception):
+    pass
+
+
+class RawDAO(Data):
+    """The so-called database access object (DAO).
+
+    The DAO models a single table in a database.
+    Specify a schema and table_name to obtain raw
+    DAO functionality, a quick and dirty query
+    builder for low-overhead pythonic database access.
+
+    Control the columns being pulled with `entities`
+    and the amount of data to be pulled with `filters`.
+    filters is a well-formed dictionary with an
+    expected schema:
+
+    .. code-block:: Python
+
+        dao = exa.core.dao.RawDAO(table_name='users')
+        # if not specified, fetches all columns
+        dao.entities = ['id', 'user']
+        # if not specified, fetches all rows
+        dao.filters = {
+            'id': [('gt', 1), ('lt', 10)],
+            'user': [('eq', 'foo')]
+        }
+        df = dao(session=sqlalchemy.Session())
+
+    """
+    schema = Unicode(help='schema name')
+    table_name = Unicode(help='table name')
+    entities = List(help='columns to pull')
+    filters = Dict(help='filters to apply')
+
+    @default('entities')
+    def _default_entities(self):
+        return self.columns
+
+    @validate('schema', 'table_name')
+    def _validate_sql(self, prop):
+        sql = prop['value']
+        self.log.debug(f"validating {sql} against sql injection")
+        if any((i in sql for i in (';', '--'))):
+            raise SQLInjectionError("; or -- found. abort")
+        return sql
+
+    def _scan_sql(self, query):
+        self.log.debug("validating query against sql injection")
+        if any((i in query for i in (';', '--'))):
+            raise SQLInjectionError("; or -- found. abort")
+
+    def _filter_builder(self, query):
+        filters = []
+        for column, conditions in self.filters.items():
+            for op, cond in conditions:
+                fil = "{} {} '{}'".format(column, _raw_op_map[op], cond)
+                self.log.debug("building filter {}".format(fil))
+                filters.append(fil)
+        filters = ' and '.join(filters)
+        if filters:
+            query += ' where {}'.format(filters)
+        return query
+
+    def _query_builder(self, query):
+        entities = ', '.join(self.entities) or '*'
+        query = 'select {} from {}'.format(entities, self.fqtn())
+        query = self._filter_builder(query)
+        self._scan_sql(query)
+        return query + ';'
+
+    def _upload(self, session, payload, chunksize=_CHUNKSIZE):
```

Comment:
Only one calls to _upload(), neither with chunksize set. Why make it configurable?

Reply:
Good point about this use of chunksize.  I think it is useful to be able to tune this value at runtime so perhaps it should just be a trait on the DAO.